### ISSUE TITLE
how to manualy define order in plotStratify() ?

### ISSUE THREAD

**Issue by guidohooiveld** [GabrielHoffman/variancePartition#46]
Hi Gabriel,
In the plot generated by `plotStratify()` (and `plotStratify()`)  I would like to manually set the order of the experimental groups, but I can't get it to work...

When using your example code from the vignette:
```
> library(variancePartition)
> data(varPartData)
> form <- ~ Age + (1|Individual) + (1|Tissue) + (1|Batch)
> varPart <- fitExtractVarPartModel( geneExpr, form, info )
Dividing work into 10 chunks...

Total:15 s
> 
> # Section 2.2 of the vignette.
> # get gene with the highest variation across Tissues
> # create data.frame with expression of gene i and Tissue
> # type for each sample
> i <- which.max( varPart$Tissue )
> GE <- data.frame( Expression = geneExpr[i,], Tissue = info$Tissue)
> 
> plotStratify( Expression ~ Tissue, GE, main=rownames(geneExpr)[i], sort=FALSE, xlab="TEST")
>
```

such 'stratified' plot is nicely made:
![image](https://user-images.githubusercontent.com/22979991/150121652-d9439699-2581-49c1-8e4f-11283d479815.png)

.... but how can I change to order of the boxes on the x-axis (and legend for 'Tissue') to e.g. B, A, C? (from A, B, C). So that the green box (B) is first, followed by the red box (A), and blue box (C).
Reason I am asking is that I am working with an experiment in which from a biological perspective there is a logic order of groups, but this order is not equal to their alphabetical order...

I had a look at your code (both for `plotStratify()` and `plotStratifyBy()`), and noticed that for both functions the order of the box plots is determined by the variable `ord`.

In the case of `plotStratify`:
https://github.com/GabrielHoffman/variancePartition/blob/45e34e9b226d38d36ad0d7385fef1f7bf0906685/R/plotStratifyBy.R#L97

By default `ord` equals `xval`:
https://github.com/GabrielHoffman/variancePartition/blob/45e34e9b226d38d36ad0d7385fef1f7bf0906685/R/plotStratifyBy.R#L93-L94

... and `xval` is set to be 'simply' the name of the 2nd column of the input data.
https://github.com/GabrielHoffman/variancePartition/blob/45e34e9b226d38d36ad0d7385fef1f7bf0906685/R/plotStratifyBy.R#L63

I played a little with this code, but I didn't get it working the way I wanted...
So since I don't how to do it, would you have a suggestion how to achieve this?
If it would not be possible by modifying the generated plot using some additional ggplot2 code/functionality, would it then be feasible to modify the function `plotStratify` so the order of groups can be set? Something like having an argument called (e.g.) `x.order`, which by default is `NULL`, but can be set? Example: `x.order <- c("B", "A", "C")`...?

Thanks for considering,
Guido

**Comment by GabrielHoffman** [comment]
The ordering is determined by the factor levels of the variable on the x-axis:

```r
# set the order uses factor levels of the x-axis variable
GE$Tissue = factor(GE$Tissue, c("B", "A", "C"))

plotStratify( Expression ~ Tissue, GE, main=rownames(geneExpr)[i], sort=TRUE, xlab="TEST")
```

**Comment by guidohooiveld** [comment]
Aha, thanks for this!
Actually, I thought along the same line before posting here. I similarly set the order of the factor levels of the metadata object `info`. After checking it again it turns out that this does the trick as well, but somehow I missed that earlier... oops!

However, one thing is not fully clear yet: how now to properly deal with / define the colors of the bars?
I am asking because after running my code below I expected A to be colored purple, B gray, and C yellow. The colors are indeed changed, but are according the **_ORDER_** they were defined, but NOT according _**HOW**_ they were defined...
For example, B=purple, but not gray...

```
> library(variancePartition)
> data(varPartData)
> form <- ~ Age + (1|Individual) + (1|Tissue) + (1|Batch)
>
> # This also works: change order of factor levels for plotting
> # already in the metadata object info.
> info$Tissue <- factor(info$Tissue, c("B", "A", "C")) 
> 
> varPart <- fitExtractVarPartModel( geneExpr, form, info )
Dividing work into 10 chunks...

Total:11 s
> 
> # Section 2.2 of the vignette.
> # get gene with the highest variation across Tissues
> # create data.frame with expression of gene i and Tissue
> # type for each sample
> i <- which.max( varPart$Tissue )
> GE <- data.frame( Expression = geneExpr[i,], Tissue = info$Tissue)
>
> #Plot order indeed B, A, C.
> plotStratify( Expression ~ Tissue, GE, main=rownames(geneExpr)[i], sort=FALSE, xlab="TEST")
>
> # Change colors of bars, and plot again.
> col = c(A="purple",  B="gray", C="yellow")
>
> plotStratify( Expression ~ Tissue, GE, main=rownames(geneExpr)[i], sort=FALSE, xlab="TEST",  colorBy=col)
> #Mmm, colors are not as manually set...

```

![image](https://user-images.githubusercontent.com/22979991/150177225-7f9cc2e1-0437-4091-a676-0a2861755bd2.png)

**Comment by GabrielHoffman** [comment]
Good point.  Fixed in `v1.25.6` in GitHub now

**Comment by guidohooiveld** [comment]
Yep, I can confirm it is working now as expected.
Thanks very much for the quick fix!
